OpenDreamRuntime/Objects/Types/DreamObjectSavefile.cs

```diff
@@ -150,31 +164,372 @@ public sealed class SavefileDirectory : Dictionary<string, DreamValue> { }
         } else {
-            _currentDirPath += path;
+            string oldPath = CurrentPath;
+            CurrentPath = new DreamPath(index).AddToPath("../").PathString; //get the parent of the target path
+            SetSavefileValue(pathArray[pathArray.Length - 1], value);
+            CurrentPath = oldPath;
+        }
+        _savefilesToFlush.Add(this);
+    }
+
+    /// <summary>
+    /// Turn the json magic value into real byond values
+    /// </summary>
+    public DreamValue DeserializeJsonValue(SFDreamJsonValue value) {
+        switch (value) {
+            case SFDreamFileValue SFDreamFileValue:
+                return new DreamValue(DreamResourceManager.CreateResource(Convert.FromBase64String(SFDreamFileValue.Data)));
+            case SFDreamListValue SFDreamListValue:
+                var l = ObjectTree.CreateList();
+                if (SFDreamListValue.AssocKeys != null) {
+                    for(int i=0; i < SFDreamListValue.AssocKeys.Count; i++) {
+                        if(SFDreamListValue.AssocData != null && SFDreamListValue.AssocData[i] != null) //note that null != DreamValue.Null
+                            l.SetValue(DeserializeJsonValue(SFDreamListValue.AssocKeys[i]), DeserializeJsonValue(SFDreamListValue.AssocData[i]!));
+                        else
+                            l.AddValue(DeserializeJsonValue(SFDreamListValue.AssocKeys[i]));
+                    }
+                }
+                return new DreamValue(l);
+            case SFDreamObjectPathValue SFDreamObjectPath:
+                SFDreamJsonValue storedObjectVars = SFDreamObjectPath;
+                SFDreamJsonValue searchDir = SFDreamObjectPath;
+                while(searchDir != _rootNode)
+                    if(!searchDir.TryGetValue(SFDreamObjectPath.Path, out storedObjectVars!))
+                        searchDir = SeekTo("..");
+                    else
+                        break;
+
+                if(storedObjectVars!.TryGetValue("type", out SFDreamJsonValue? storedObjectTypeJSON) && DeserializeJsonValue(storedObjectTypeJSON).TryGetValueAsType(out TreeEntry? objecTypeActual)) {
+                    DreamObject resultObj = _objectTree.CreateObject(objecTypeActual);
+                    foreach(string key in storedObjectVars.Keys){
+                        if(key == "type")
+                            continue;
+                        resultObj.SetVariable(key, DeserializeJsonValue(storedObjectVars[key]));
+                    }
+                    resultObj.InitSpawn(new DreamProcArguments());
+                    resultObj.SpawnProc("Read", null, [new DreamValue(this)]);
+                    return new DreamValue(resultObj);
+                } else
+                    throw new InvalidDataException("Unable to deserialize object in savefile: " + ((storedObjectTypeJSON as SFDreamType) is null ? "no type specified (corrupted savefile?)" : "invalid type "+((SFDreamType)storedObjectTypeJSON!).TypePath));
```

Comment:
silently errors apparently
https://discord.com/channels/847318834952405042/887074737879138315/1200283136026755152

Reply:
well that's insane. How the hell is it inferring the type if type is overridden?
Honestly think this one should be an error.

Reply:
Turns out the test case was bad, `F << S["foo"]` is a no-op. BYOND just silently fails and outputs a null. We choose to fail loudly instead.